Update PGA doc for UAST dataset (#170)
src-d/datasets: PublicGitArchiveUASTs/README.md

UASTs extracted from Public Git Archive ![size 5TB](https://img.shields.io/badge/size-5TB-green.svg)
=======================================

The [Universal Abstract Syntax Trees](https://doc.bblf.sh/uast/uast-specification-v2.html) (UASTs) extracted
from the latest (HEAD) revision of every Git reference contained in [Public Git Archive](../../PublicGitArchive).
The dataset is distributed as Parquet files, which you can download using the [pga CLI](../PublicGitArchive/pga).
There is also a [ClickHouse DB version](ClickHouse) which is more lightweight and easier to work with.

### Format

> The Parquet files have 3 columns (each row is a single file):

Author: done

- `head` (string): the UUID of the repository of the given file. You can use [this mapping](https://drive.google.com/open?id=136vsGWfIwfd0IrAdfphIU6lkMmme4-Pj)
to obtain the repository names from UUIDs.
- `path` (string): the filepath to the given file, in the repository structure.
- `uast` (variable-length byte array): the UAST of the given file.

### Usage

> The Parquet files can be read using any library that supports the format, however using Spark is strongly advised if processing a large part of the dataset. The UASTs are stored as byte arrays, and thus you can use any of the [Babelfish Clients](https://doc.bblf.sh/using-babelfish/clients.html) to read and manipulate them.

Author: done

> As an example, here is how one would extract all identifiers from the UASTs in a given Parquet file:

Reviewer: ```suggestion
For example, this is how to extract all identifiers from the UASTs in a given Parquet file:
```

Author: done

> ```

Author: done

# We create the Spark Config - tune accordingly

conf = SparkConf().setAll([ ... ])

# We create the Spark Session - the master URL may be wrong depending on your cluster

spark = SparkSession.builder \
    .appName("pga-example") \
    .master("spark://spark-spark-master:7077") \
    .config(conf=conf) \
    .getOrCreate()

# We define the function that will extract identifiers from each UAST

def extract_identifiers(uast):
    ctx = bblfsh.decode(uast)  # Decode the Byte Array and create the Context
    identifiers = []
    for node in ctx.filter("//uast:Identifier"):  # Iterate over the identifier nodes
        node = node.load()  # Load the node in memory
        identifiers.append(node["Name"])  # Extract the identifier from the node
    return identifiers

# We create the Spark User Defined Function usaing above function

extract_identifiers_udf = udf(extract_identifiers, ArrayType(StringType()))

# We apply the pipeline, then trigger execution with `show`

df = spark.read.parquet("/path/to/parquet")
df = df.withColumn("identifier", explode(extract_identifiers_udf(df.uast))) \
    .select("head", "path", "identifier")
df.show()
```

> Please note that the [Babelfish Python Client](https://github.com/bblfsh/python-client) needs to be present on the Spark workers for this snippet to work, **not only on the driver.**

Author: done

### Origin

We used [pga2uast](../PublicGitArchive/pga2uast) to parse the files in HEAD revisions of PGA repositories.
Please refer to [this GitHub issue](https://github.com/src-d/ml-backlog/issues/74) that describes
the procedure in high detail. It was quite sophisticated because we wanted to cover as much data as we could.
We used 11 "Start-2-L" machines on online.net.

### Limitations

|           | # of repos | # of files | # of distinct files | % of duplicates |
|:---------:|:----------:|:----------:|:-------------------:|:---------------:|
| **PGA**   | 220,174    | 40,971,787 | 40,829,244          | 0.3 %           |
| **UASTs** | 218,023    | 36,162,330 | 35,991,340          | 0.5 %           |

> As the table above shows, we were not able to process 100% of the HEAD of Public Git Archive. For one, we could not process all languages, as Babelfish currently only has drivers for 9 languages. Furthermore, some files proved too large to be processable in a reasonnable amount of time. This, combined to parsing errors or bugs on Babelfish's side, resulted in missing ~12% of all parsable files in the HEAD of PGA, amounting for ~45% of the data in bytesizes. As we can see from the table below, not all languages all equal: for instance, the C++ driver, which handles all C-like languages (C, C++, Metal, Cuda), performed much worse then most lamguages, while the Go driver performed much better then the rest of the languages.

Author: done
